### PR TITLE
updates ordering test scripts and adds a few more handlers to keep ordering intact

### PR DESCRIPTION
**Pull request by akshaydeo**
## Summary

Fixes JSON key ordering preservation and serialization fidelity issues in Bifrost's request proxying pipeline. Addresses three critical problems: tool_choice struct serialization injecting zero-value fields and reordering keys, tool parameter properties losing their original key order, and $defs definitions moving from their intended position within schema objects.

## Changes

- **Added `MergeExtraParamsIntoJSON` function**: Preserves original JSON key ordering when merging extra parameters by parsing token-by-token instead of round-tripping through `map[string]interface{}`
- **Implemented custom `MarshalJSON`/`UnmarshalJSON` for `ChatToolChoiceStruct`**: Ensures only active variant fields are serialized with consistent "type"-first ordering, preventing zero-value field injection
- **Enhanced `OrderedMap` with proper JSON serialization**: Maintains key insertion order for tool function parameters and schema definitions
- **Switched raw request/response storage to `json.RawMessage`**: Preserves exact JSON bytes including key ordering instead of parsing into generic interfaces
- **Removed concurrent unmarshaling**: Eliminated goroutines in `HandleProviderResponse` that were unnecessary for the simplified `json.RawMessage` approach
- **Added comprehensive test coverage**: Load test with 8 payload variants covering edge cases like reverse-alphabetical properties, deeply nested objects, and adversarial key orderings

## Type of change

- [x] Bug fix
- [ ] Feature
- [ ] Refactor
- [ ] Documentation
- [ ] Chore/CI

## Affected areas

- [x] Core (Go)
- [ ] Transports (HTTP)
- [x] Providers/Integrations
- [ ] Plugins
- [ ] UI (Next.js)
- [ ] Docs

## How to test

Validate JSON key ordering preservation across different payload structures:

```sh
# Run the comprehensive load test
./tests/load_test_parameter_ordering.sh

# Run unit tests for the new functionality
go test ./core/providers/utils -v -run TestMergeExtraParamsIntoJSON
go test ./core/schemas -v -run TestSonic_ChatToolChoiceStruct

# Test with a real request to verify tool_choice serialization
curl -X POST http://localhost:8080/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d '{
    "model": "openai/gpt-4",
    "messages": [{"role": "user", "content": "test"}],
    "tools": [{"type": "function", "function": {"name": "test", "parameters": {"type": "object", "properties": {"answer": {"type": "string"}, "reasoning": {"type": "string"}}}}}],
    "tool_choice": {"type": "function", "function": {"name": "test"}}
  }'
```

The response should not contain extra fields like `"custom":{"name":""}` or `"allowed_tools":null` in the tool_choice object.

## Screenshots/Recordings

N/A - Backend JSON serialization changes

## Breaking changes

- [ ] Yes
- [x] No

This is a bug fix that preserves existing API behavior while fixing serialization fidelity issues.

## Related issues

Fixes JSON key ordering mutations in tool parameter proxying and tool_choice serialization that were causing schema validation failures and inconsistent request formatting.

## Security considerations

No security implications. Changes only affect JSON serialization fidelity and do not modify authentication, authorization, or data handling logic.

## Checklist

- [x] I read `docs/contributing/README.md` and followed the guidelines
- [x] I added/updated tests where appropriate
- [x] I updated documentation where needed
- [x] I verified builds succeed (Go and UI)
- [x] I verified the CI pipeline passes locally if applicable

### CODE REVIEW

**Comment by akshaydeo** [comment]
* **#1876** <a href="https://app.graphite.com/github/pr/maximhq/bifrost/1876?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/maximhq/bifrost/1876?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: core/providers/utils/utils.go [hunk 1]
```diff
@@ -645,25 +739,15 @@ func EnrichError(
...
 
 	if ShouldSendBackRawRequest(ctx, sendBackRawRequest) && len(requestBody) > 0 {
-		var rawRequest interface{}
-		if err := sonic.Unmarshal(requestBody, &rawRequest); err != nil {
-			getLogger().Warn("Failed to parse raw request for error: %v", err)
-			return bifrostErr
-		}
-		bifrostErr.ExtraFields.RawRequest = rawRequest
+		// Store as json.RawMessage to preserve exact JSON bytes (including key ordering)
+		bifrostErr.ExtraFields.RawRequest = json.RawMessage(requestBody)
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/utils/utils.go:844]: This is not a concern in practice. The byte slices are already safe copies by the time they reach `EnrichError` and `HandleProviderResponse`:

1. **Response body**: `CheckAndDecodeBody()` explicitly allocates a new byte slice and copies the data (`make([]byte, len(body))` + `copy(result, body)`), so `responseBody` is never a reference to fasthttp's pooled buffer.
2. **Request body**: Comes from either `sonic.MarshalIndent` (new allocation) or `request.GetRawRequestBody()` (backed by `RequestCtx` which stays valid for the entire request lifecycle).

Both functions are called before any fasthttp buffer is released, so there's no use-after-free risk.

File: core/schemas/chatcompletions.go [hunk 2]
```diff
@@ -503,6 +503,83 @@ type ChatToolChoiceStruct struct {
...
+	case ChatToolChoiceTypeCustom:
+		s.Function = nil
+		s.AllowedTools = nil
+	case ChatToolChoiceTypeAllowedTools:
+		s.Function = nil
+		s.Custom = nil
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/schemas/chatcompletions.go:584]: The existing switch already handles the important case — cleaning up sonic's zero-value allocations for non-active variants. If the active variant key is absent (e.g., `{"type":"function"}` with no `"function"` key), the pointer will already be `nil` (sonic doesn't allocate zero-value pointers for truly absent fields when using the Alias trick). Adding an extra `json.Unmarshal` into `map[string]json.RawMessage` for every unmarshal call would add overhead without practical benefit — the caller already needs to handle `nil` active variant pointers regardless.

File: core/providers/utils/utils.go [hunk 3]
```diff
@@ -457,6 +458,106 @@ func MergeExtraParams(jsonMap map[string]interface{}, extraParams map[string]int
...
+					}
+				}
+			}
+			// Non-object or merge failed: overwrite in place (preserving position)
+			pairs[idx].val = newValBytes
+		} else {
+			// New key: append at end
+			pairs = append(pairs, kvPair{k, newValBytes})
+		}
+	}
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/utils/utils.go]: Fixed — sorted extraParams keys before iterating to ensure deterministic append order for new keys.

File: core/providers/utils/utils.go [hunk 4]
```diff
@@ -457,6 +458,106 @@ func MergeExtraParams(jsonMap map[string]interface{}, extraParams map[string]int
...
+			existingTrimmed := bytes.TrimSpace(pairs[idx].val)
+			newTrimmed := bytes.TrimSpace(newValBytes)
+			if len(existingTrimmed) > 0 && existingTrimmed[0] == '{' &&
+				len(newTrimmed) > 0 && newTrimmed[0] == '{' {
+				var existingMap, newMap map[string]interface{}
+				if json.Unmarshal(existingTrimmed, &existingMap) == nil {
+					if json.Unmarshal(newTrimmed, &newMap) == nil {
+						MergeExtraParams(existingMap, newMap)
+						if merged, err := sonic.Marshal(existingMap); err == nil {
+							pairs[idx].val = merged
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply core/providers/utils/utils.go:622]: Fixed — nested merge now uses `json.NewDecoder` with `UseNumber()` to preserve numeric precision during the round-trip.

File: helm-charts/index.yaml [hunk 5]
```diff
@@ -1,6 +1,27 @@
...
+    maintainers:
+    - email: akshay@getmaxim.ai
+      name: Bifrost Team
+    name: bifrost
+    sources:
+    - https://github.com/maximhq/bifrost
+    type: application
+    urls:
+    - https://maximhq.github.io/bifrost/helm-charts/bifrost-2.0.10.tgz
+    version: 2.0.10
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply helm-charts/index.yaml:24]: Fixed — updated `generated` timestamp to match the new entry's `created` timestamp.

File: tests/load_test_parameter_ordering.sh [hunk 6]
```diff
@@ -142,32 +465,120 @@ def extract_key_orders(obj, path=""):
...
+      - Zero-value fields injected: "custom":{"name":""}, "allowed_tools":{"mode":"","tools":null}
+      - Key reordering: "type" moving from first to last position
+      - String tool_choice ("auto"/"none"/"required") being expanded to struct
+    """
+    mutations = []
+    input_tc = input_payload.get("tool_choice")
+    output_tc = raw_request.get("tool_choice")
+    if input_tc is None or output_tc is None:
+        return mutations
+
```
[the remark this answers is not in the document]
Comment (akshaydeo) [reply tests/load_test_parameter_ordering.sh]: Fixed — now detects one-sided tool_choice mutations (injected/dropped) instead of silently skipping them.
